RAM disk

erikng opened the pull request:
### What does this PR do?
This adds an option to use a RAM disk with the `image` component.

### New Behavior
Add the following to your config.plist image component

```xml
          <key>ramdisk</key>
          <true/>
```

Imagr will get the `Content-Length` of the dmg file and if there is enough available RAM, it will create a RAM disk and use it, otherwise it will just fallback to asr. If urllib2 cannot validate SSL (say connecting to TLS 1.2 on a pre 10.13 NBI), it will also fallback to asr.

This allows us to utilize gurl to download the DMG vs asr. gurl is significantly more performant in both local and CDN downloads, allowing a faster imaging proceess.

erikng committed:
ramdisk working via Imagr.app in Booted OS

erikng committed:
show progress for download

grahamgilbert commented:
Two things:

* This shouldn’t be the default. It looks like you are passing true as the default. 
* This should be a randomly generated volume name - or at the very least there should be a check for existing volumes of the same name. I would imagine this is a common name that would be used in various scripts

erikng commented:
Which part is making it default? The function passes false. 

Okay I’ll look into the random volume name vs hardcoding ramdisk.

gregneagle commented:
> Add the following to your config.plist image component
> 
>           <key>ramdisk</key>
>           <true/>
> Or optionally add a ramdiskbytes key if you have a bigger image to deploy.

Meaning if you add `ramdiskbytes` you don't need to add `ramdisk`? If so, what happens if someone does add both?

gregneagle commented:
After looking at the code, this should be worded:

> Add the following to your config.plist image component
> 
>           <key>ramdisk</key>
>           <true/>
>
> Optionally also add a ramdiskbytes key if you have a bigger image to deploy; the default size is XXXX bytes.
> 
>           <key>ramdiskbytes</key>
>           <string>7516192768</string>

grahamgilbert commented:
Following on from @keeleysam’s comments on Slack, what about testing the size of the dmg and the amount of ram and do the right thing automagically?

erikng committed:
add getDMGSize function to utils.py

erikng committed:
remove ramdiskbytes and default to no ramdisk

erikng committed:
remove ramdiskbytes from clone function

erikng committed:
random ramdisk volume name and logic for image size

erikng committed:
ensure filesize is an int

erikng commented:
Changes are made and tested on 10.12 and 10.13 NBIs.

grahamgilbert commented:
Just tried on a 10.13 AutoNBI generated disk image and I get a hard crash

```
2017-10-16 19:08:35.812 Imagr[668:95967] NSColorList could not parse color list file /System/Library/Colors/System.clr
Keyboard Layouts: duplicate keyboard layout identifier -14934.
Keyboard Layouts: keyboard layout identifier -14934 has been replaced with -28673.
2017-10-16 19:08:35.989 Imagr[668:95967] Application did finish launching.
2017-10-16 19:08:35.991 Imagr[668:95967] background_window is set to u'auto'
2017-10-16 19:08:44.693 Imagr[668:96022] Report: https://s3.amazonaws.com/ourimagrbucket/imagr_config.plist
2017-10-16 19:08:44.696 Imagr[668:96022] gurl options: {
    "additional_headers" = "<null>";
    file = "/tmp/tmpkENHcj/tempdata.download";
    "follow_redirects" = 0;
    "logging_function" = "<function NSLogWrapper at 0x111de01b8>";
    password = "<null>";
    url = "https://s3.amazonaws.com/ourimagrbucket/imagr_config.plist";
    username = "<null>";
}
2017-10-16 19:08:44.863 Imagr[668:96022] connection_willSendRequestForAuthenticationChallenge_
2017-10-16 19:08:44.863 Imagr[668:96022] Authentication challenge for Host: s3.amazonaws.com Realm: None AuthMethod: NSURLAuthenticationMethodServerTrust
2017-10-16 19:08:44.864 Imagr[668:96022] Allowing OS to handle authentication request
2017-10-16 19:08:45.121 Imagr[668:96022] Percent done: 100
2017-10-16 19:08:45.121 Imagr[668:96022] Status: 200
2017-10-16 19:08:45.121 Imagr[668:96022] Headers: {
    "Accept-Ranges" = bytes;
    "Content-Length" = 30668;
    "Content-Type" = "binary/octet-stream";
    Date = "Mon, 16 Oct 2017 19:08:46 GMT";
    Etag = "\"0417ef5076755a41b4c50dabfed7e0c4\"";
    "Last-Modified" = "Mon, 16 Oct 2017 18:30:34 GMT";
    Server = AmazonS3;
    "x-amz-id-2" = "mfYuLw+ecwfnmzGlZG1SCYnlf+mBdZTX9ybXfCv0afw4lXTk7AMI7ti89eY9ARVG7TVM6RyQMdg=";
    "x-amz-request-id" = 0F01602B0382FBCC;
    "x-amz-version-id" = pRWJxqS5V01Xu81P8esy5NYw7eDQVEEW;
}
2017-10-16 19:08:45.125 Imagr[668:96852] Loading background image
2017-10-16 19:08:45.151 Imagr[668:95967] Showing background window
2017-10-16 19:08:52.730 Imagr[668:96871] gurl options: {
    "additional_headers" = "<null>";
    file = "/tmp/tmpdEZ2Xb/tempdata.download";
    "follow_redirects" = 0;
    "logging_function" = "<function NSLogWrapper at 0x111de01b8>";
    password = "<null>";
    url = "https://s3.amazonaws.com/ourimagrbucket/scripts/lr_inventory_check.sh";
    username = "<null>";
}
2017-10-16 19:08:52.875 Imagr[668:96871] connection_willSendRequestForAuthenticationChallenge_
2017-10-16 19:08:52.875 Imagr[668:96871] Authentication challenge for Host: s3.amazonaws.com Realm: None AuthMethod: NSURLAuthenticationMethodServerTrust
2017-10-16 19:08:52.875 Imagr[668:96871] Allowing OS to handle authentication request
2017-10-16 19:08:53.098 Imagr[668:96871] Percent done: 100
2017-10-16 19:08:53.098 Imagr[668:96871] Status: 200
2017-10-16 19:08:53.098 Imagr[668:96871] Headers: {
    "Accept-Ranges" = bytes;
    "Content-Length" = 437;
    "Content-Type" = "application/x-sh";
    Date = "Mon, 16 Oct 2017 19:08:54 GMT";
    Etag = "\"767aa26bba164b9969bcb189102df047\"";
    "Last-Modified" = "Tue, 12 Sep 2017 22:16:56 GMT";
    Server = AmazonS3;
    "x-amz-id-2" = "AmVdwocMJSAEn1yTL6XS9pwN+rZd0BbKTmZ9ryuqpGtg+lgMuF9WNESDeVLQfVTKJ1ixYa05Mdo=";
    "x-amz-request-id" = 12BBF63202E1D23E;
    "x-amz-version-id" = KGkm9HmT1H5QY2sIbNr2M1UcuJREQW1a;
}
2017-10-16 19:08:55.648 Imagr[668:96871] Total Memory is 17179869184
2017-10-16 19:08:55.648 Imagr[668:96871] Available Memory for image is 16655581184
2017-10-16 19:08:56.185 Imagr[668:96871] Required Memory for image is 8959145391
2017-10-16 19:08:56.185 Imagr[668:96871] Amount of Sectors for RAM Disk is 17498330
2017-10-16 19:09:07.300 Imagr[668:96871] Result of isApfs is True
2017-10-16 19:09:07.300 Imagr[668:96871] Formatting RAM Disk as APFS at /dev/disk24
2017-10-16 19:09:07.312 Imagr[668:96871] Mounting APFS RAM Disk /dev/disk24
2017-10-16 19:09:07.322 Imagr[668:96871] Downloading DMG file from https://s3.amazonaws.com/ourimagrbucket/masters/highsierra.dmg
2017-10-16 19:09:07.323 Imagr[668:96871] gurl options: {
    "additional_headers" = "<null>";
    file = "/Volumes/ramdisk7326248/highsierra.dmg.download";
    "follow_redirects" = 0;
    "logging_function" = "<function NSLogWrapper at 0x111de01b8>";
    password = "<null>";
    url = "https://s3.amazonaws.com/ourimagrbucket/masters/highsierra.dmg";
    username = "<null>";
}
2017-10-16 19:09:07.466 Imagr[668:96871] connection_willSendRequestForAuthenticationChallenge_
2017-10-16 19:09:07.466 Imagr[668:96871] Authentication challenge for Host: s3.amazonaws.com Realm: None AuthMethod: NSURLAuthenticationMethodServerTrust
2017-10-16 19:09:07.466 Imagr[668:96871] Allowing OS to handle authentication request
2017-10-16 19:09:07.706 Imagr[668:96871] *** Terminating app due to uncaught exception 'OC_PythonException', reason: '<type 'exceptions.IOError'>: [Errno 2] No such file or directory: u'/Volumes/ramdisk7326248/highsierra.dmg.download''
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000103a9a32b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x0000000101b3a942 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000103a9a279 -[NSException raise] + 9
	3   _objc.so                            0x000000010fbe03d1 PyObjCErr_ToObjCWithGILState + 46
	4   _objc.so                            0x000000010fbcd94e PyObjCFFI_MakeClosure + 5169
	5   libffi.dylib                        0x00000001042929f3 ffi_closure_unix64_inner + 487
	6   libffi.dylib                        0x00000001042920e6 ffi_closure_unix64 + 70
	7   CFNetwork                           0x000000010854e4ad -[NSURLConnectionInternal _withConnectionAndDelegate:onlyActive:] + 197
	8   CFNetwork                           0x000000010854e3d6 -[NSURLConnectionInternal _withActiveConnectionAndDelegate:] + 48
	9   CFNetwork                           0x0000000108551d74 _ZL34_NSURLConnectionDidReceiveResponseP16_CFURLConnectionP14_CFURLResponsePKv + 81
	10  CFNetwork                           0x0000000108551c8e ___ZN27URLConnectionClient_Classic28_delegate_didReceiveResponseEP14_CFURLResponse_block_invoke + 110
	11  CFNetwork                           0x00000001086bc913 ___ZN27URLConnectionClient_Classic18_withDelegateAsyncEPKcU13block_pointerFvP16_CFURLConnectionPK33CFURLConnectionClientCurrent_VMaxE_block_invoke_2 + 100
	12  libdispatch.dylib                   0x00000001041e1f64 _dispatch_client_callout + 8
	13  libdispatch.dylib                   0x00000001041f454d _dispatch_block_invoke_direct + 317
	14  CFNetwork                           0x000000010854e2a8 _ZN19RunloopBlockContext13_invoke_blockEPKvPv + 24
	15  CoreFoundation                      0x00000001039da428 CFArrayApplyFunction + 72
	16  CFNetwork                           0x000000010854e17e _ZN19RunloopBlockContext7performEv + 132
	17  CFNetwork                           0x000000010854e01c _ZN17MultiplexerSource7performEv + 282
	18  CFNetwork                           0x000000010854de4f _ZN17MultiplexerSource8_performEPv + 65
	19  CoreFoundation                      0x0000000103a33941 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	20  CoreFoundation                      0x0000000103aeb33c __CFRunLoopDoSource0 + 108
	21  CoreFoundation                      0x0000000103a16930 __CFRunLoopDoSources0 + 208
	22  CoreFoundation                      0x0000000103a15dad __CFRunLoopRun + 1293
	23  CoreFoundation                      0x0000000103a15607 CFRunLoopRunSpecific + 487
	24  Foundation                          0x00000001014733f6 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 277
	25  Foundation                          0x00000001014f96b5 -[NSRunLoop(NSRunLoop) runUntilDate:] + 114
	26  libffi.dylib                        0x0000000104291f64 ffi_call_unix64 + 76
	27  ???                                 0x00007fc644f68090 0x0 + 140489537257616
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

erikng committed:
fix APFS ram disk mount command and sleep

diskuitl mount /dev/diskXX does not work for APFS volumes.

diskutil erasedisk TYPE VOLUMENAME DISKPATH works for both APFS and HFS+. This command properly initializes/erases/mounts the disk.

Sleep 2 seconds, because even though the process is finished for initializing the disk, it's actually not fully initialized. Prior to this, gurl was being invoked 2 tenths of a second after the subprocess was finished, resulting in disk IO errors.

erikng committed:
bump netinstall RAM requirement to 1GB

erikng committed:
Pad the needed RAM disk filesize due to formatting loss

erikng committed:
retry for failed gurl timeouts

erikng committed:
if target volume doesnt match image dont create RAMDisk or download image

erikng committed:
Properly bail if three dmg download trys fail.

erikng committed:
pass can_resume to gurl for DMG downloads

erikng committed:
detach RAM disk if there is a download failure.

erikng commented:
I've added some features:

* Retry the download three times in case of network drop and bail if it doesn't work.
* pass `can_resume` to gurl for DMG downloads.
* Don't attempt to setup the RAM disk/download the DMG if the source/target volume don't match.
* Detach the RAM disk if it fails, so user can try again without potentially locking up the machine.

erikng committed:
remove error, causing Imagr to falsely fail on gurl retries

erikng committed:
detach RAMdisk after successful image

erikng committed:
lower RecoveryOS to 650MB and 6% disk utilization padding

RecoveryOS really uses about 650MB of RAM, and 6% padding is sufficient for RAMdisk.

erikng committed:
do not use RAMdisk if under 8GB of RAM

erikng commented:
I have tested this with the following configurations this morning:

with <ramdisk> key
1. VM with 8192 MB of RAM. This resulted in ASR imaging.
2. VM with 9216 MB of RAM. This resulted in RAM disk imaging.
3. VM with 8192 MB of RAM inside an included workflow. This resulted in ASR imaging.
4. VM with 9216 MB of RAM inside an included workflow. This resulted in RAM disk imaging.
5. MBA with 8192 MB of RAM. This resulted in ASR imaging.
6. VM with 9216 MB of RAM from a local http URL.
6. VM with 9216 MB of RAM from a CDN based https URL.

without <ramdisk> key
1. VM with 8192 MB of RAM. This resulted in ASR imaging.
2. VM with 9216 MB of RAM. This resulted in ASR imaging.

Clkw0rk commented:
I've tested this with under the following conditions:

with `ramdisk` key as well as an AutdoDMG created `10.12.6.hfs.dmg`:
10.12.6 dmg is `6864954159 bytes`

1. MacBookPro with 16GB RAM - RAM disk Imaging from local http share
2. MacBookPro with 8GB RAM - RAM disk Imaging from local http share
3. Mac mini with 8GB RAM - RAM disk Imaging from local http share

I've put these machines through the workflow 2-3 times each now without running into any issues.

erikng committed:
Merge branch 'master' of https://github.com/grahamgilbert/imagr into ramdisk

erikng commented:
I'm moving this to a function so it can potentially be used with startosinstall. Currently re-testing.

erikng committed:
move RAMDisk to a function

erikng commented:
Retested with similar setups to the one from last week and it's working as expected.

erikng committed:
support for startosinstall

erikng commented:
Tested the following with `startosinstall`:

with `ramdisk` key
1. VM with 8GB of RAM - ran RAM disk startosinstall
2, VM with 4GB of RAM - ran normal startosinstall
3. VM with 8GB of RAM, but disable network midway through - ran RAM disk startosinstall and attempted to rerun gurl 3 times. If network did not recover, it correctly exited out with a posted msg/volume refresh.

with `ramdisk` key and additional packages in workflow
1. VM with 8GB of RAM - ran RAM disk startosinstall and additional packages

without `ramdisk` key
1. VM - ran normal startosinstall

There is a bug that I have found though and one I am hoping @gregneagle can help with.

When attempting to download a DMG with spaces, I get the following error from gurl:
`Gurl Error: GurlError(-1, 'Too few arguments for format string [cur:1/len:1]')`

working
```xml
        <dict>
          <key>type</key>
          <string>startosinstall</string>
          <key>ramdisk</key>
          <true/>
          <key>url</key>
          <string>http://10.10.10.10/imagr/dmgs/Install%20macOS%20High%20Sierra-10.13.0.dmg</string>
        </dict>
```

not working
```xml
        <dict>
          <key>type</key>
          <string>startosinstall</string>
          <key>ramdisk</key>
          <true/>
          <key>url</key>
          <string>http://10.10.10.10/imagr/dmgs/InstallmacOSHighSierra-10.13.0.dmg</string>
        </dict>
```

gregneagle commented:
Does this downloading bug also happen with packages that have spaces? It's probably not a gurl bug since Munki uses gurl, and I definitely have packages in my Munki repo with spaces in their name:
```xml
	<key>display_name</key>
	<string>Install macOS High Sierra</string>
	<key>installed_size</key>
	<integer>9227469</integer>
	<key>installer_item_hash</key>
	<string>8cec5756b64d2e7a394bd5fa8299ef2d72133afeee9ae959ddf65ac543b610fd</string>
	<key>installer_item_location</key>
	<string>os_updates/Install macOS High Sierra-10.13.dmg</string>
	<key>installer_item_size</key>
	<integer>5053038</integer>
	<key>installer_type</key>
	<string>startosinstall</string>
	<key>minimum_munki_version</key>
	<string>3.0.0.3211</string>
	<key>minimum_os_version</key>
	<string>10.8</string>
	<key>name</key>
	<string>InstallHighSierra</string>
```

gregneagle commented on Imagr/MainController.py:
```diff
@@ -1105,17 +1120,153 @@ def Clone(self, source, target, erase=True, verify=True, show_activity=True):
+            ramdiskvolname = "ramdisk" + str(randomnum)
+            NSLog(u"RAM Disk mountpoint is %@", str(ramdiskvolname))
+            # Only check for apfs_image status if imaging, startosinstall
+            # should just use apfs since it's meant for 10.13 and higher.
```
`startosinstall` doesn't format a disk image at all; all the disk images containing "Install macOS High Sierra.app" I've used are HFS+ formatted, which would be important if you were trying to upgrade a Mac running macOS < 10.12.x.  I'm not convinced of the need to format your ramdisk as anything other than HFS+.

erikng replied:
The only reason I am choosing this is because APFS = future and HFS could potentially go away in a future NBI. I'd rather not have to fix this later if that's the case.

gregneagle replied:
Pretty sure every single ramdisk that gets mounted in the macOS Installer environment (even 10.13) is formatted HFS+. I don't think HFS+ support will go away from macOS for a very long time. But your call.

erikng replied:
In your opinion, should `startosinstall` (in an Imagr context, not in the fact it can work) support 10.11 10.12 NBIs? If so, then I need probably need to do HFS.

gregneagle commented on Imagr/MainController.py:
```diff
@@ -1179,6 +1330,31 @@ def downloadAndInstallPackage(self, url, target, progress_method=None, additiona
+        if not os.path.basename(url).endswith('.dmg'):
+            self.errorMessage = "%s doesn't end with either '.dmg'" % url
+            return False
+        if os.path.basename(url).endswith('.dmg'):
```
This test is unnecessary: you already returned False if the os.path.basename(url).endswith('.dmg') doesn't end with '.dmg'

erikng replied:
rewrote

gregneagle commented on Imagr/Utils.py:
```diff
@@ -354,6 +355,29 @@ def downloadFile(url, additional_headers=None, username=None, password=None):
+        try:
+            dmg = urllib2.urlopen(request)
+            data = dmg.headers['Content-Length']
+        except urllib2.URLError, err:
```
These two except clauses could be combined:
```python
    except (urllib2.URLError, urllib2.HTTPError), err:
        setattr(error, 'reason', err)
        data = False
```

erikng replied:
fixed

erikng committed:
change dmg validation

erikng committed:
combine except clause

gregneagle commented on Imagr/MainController.py:
```diff
@@ -1179,6 +1330,31 @@ def downloadAndInstallPackage(self, url, target, progress_method=None, additiona
+            if failsleft == 0:
+                return False
+        else:
+            self.errorMessage = "%s doesn't end with either '.dmg'" % url
```
"either" is misplaced here.

erikng replied:
fixed

erikng committed:
remove either

gregneagle commented:
As for your issue with spaces -- there are so many layers on top of gurl that I'd be tempted to add an NSLog statement to each one to print out the url at each stage to make sure gurl is getting the "raw" url and not one that has been altered.

gregneagle commented:
gurl is called by Utils.get_url, which is called by Utils.downloadFile and Utils.downloadChunks;

Utils.downloadChunks is called by MainController.downloadPackage, MainController.downloadAndInstallPackage and some of your new code...

Utils.downloadFile is called by MainController.loadData, MainController.runComponent

erikng commented:
Some kind of encoding must be occuring.

gurl can handle %20 when pointed at the imagr config.plist file.

erikng committed:
NSLog was breaking URLs with spaces when using gurl

erikng commented:
@gregneagle helped me figure out the %20 issue.

It appears that I found an Imagr bug that impacted packages/dmgs as well, but either no one had noticed it or no one was using spaces/gurl.

erikng committed:
wait until the ramdisk is fully mounted

erikng committed:
pull apfs/hfs formatting of RAM disks

erikng committed:
bump back to 7% ramdisk padding

erikng committed:
bump to 10% disk padding

erikng committed:
Merge branch 'master' of https://github.com/grahamgilbert/imagr into ramdisk
